### PR TITLE
Add vdpa interface support

### PR DESCRIPTION
**Pull request by makoto126**
**What this PR does / why we need it**:
Added a VDPA interface type to allow creation of virtual machines with vdpa NICs. VDPA devices are usually allocated by [Intel SR-IOV device plugin](https://github.com/intel/sriov-network-device-plugin). 
Examples:
```
          interfaces:
          - masquerade: {}
            name: default
          - name: offload
            vdpa: {}
      networks:
      - name: default
        pod: {}
      - multus:
          networkName: default/offload
        name: offload
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
 [Intel SR-IOV device plugin](https://github.com/intel/sriov-network-device-plugin) for vdpa devices currently has a [bug](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pull/493) that has not been merged. It works fine with kubevirt after I merged it locally. I think it will be merged soon.

**Release note**:
```release-note
Allow creation of virtual machines with vdpa NICs.
```

### CODE REVIEW

File: pkg/util/util.go [hunk 1]
```diff
@@ -53,6 +53,15 @@ func IsSRIOVVmi(vmi *v1.VirtualMachineInstance) bool {
 	return false
 }
 
+func IsVDPAVMI(vmi *v1.VirtualMachineInstance) bool {
```
Comment (prnaraya) [inline pkg/util/util.go:56]: Nit: VDPAVMI is a bit difficult to read/understand right away. Following the convention of the above func `IsSRIOVVmi` would be a little easier to read.
```suggestion
func IsVDPAVmi(vmi *v1.VirtualMachineInstance) bool {
```

File: pkg/virt-handler/isolation/detector.go [hunk 2]
```diff
@@ -156,7 +156,7 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
 // virt-launcher pod on the given VMI according to its spec.
 // Only VMI's with VFIO devices (e.g: SRIOV, GPU), SEV or RealTime workloads require QEMU process MEMLOCK adjustment.
```
Comment (prnaraya) [inline pkg/virt-handler/isolation/detector.go:157]: Do vDPA devices require it as well? I would update this comment with that information.

File: pkg/virt-launcher/virtwrap/converter/network.go [hunk 3]
```diff
@@ -149,6 +162,53 @@ func CreateDomainInterfaces(vmi *v1.VirtualMachineInstance, domain *api.Domain,
...
+	})
+	for _, iface := range ifaces {
+		resourceEnvVarName := fmt.Sprintf("KUBEVIRT_RESOURCE_NAME_%s", iface.Name)
+		resource, isSet := os.LookupEnv(resourceEnvVarName)
+		if !isSet {
+			log.Log.Warningf("%s not set for VDPA interface %s", resourceEnvVarName, iface.Name)
+			continue
+		}
+		resources = append(resources, resource)
+		for _, resource := range resources {
```
Comment (prnaraya) [inline pkg/virt-launcher/virtwrap/converter/network.go:180]: `resource` is already defined and doesn't need to be defined again here.
```suggestion
		for _, resource = range resources {
```

**Review by prnaraya — COMMENTED** [review]
Thanks for the PR, I've added some comments inline. Additionally, is this PR then blocked by the [PR](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pull/493) you linked? If so, I would explicitly state that in the special notes for your reviewer section.

File: pkg/virt-handler/isolation/detector.go (hunk 2 above)
Comment (makoto126) [reply pkg/virt-handler/isolation/detector.go:157]: Yes, vDPA devices also needs to adjust the memlock limit. Take a look at [this](https://patchew.org/Libvirt/20210323185052.961910-1-laine@redhat.com/).

**Comment by makoto126** [comment]
@prnaraya If using [Intel SR-IOV device plugin](https://github.com/intel/sriov-network-device-plugin) for vdpa devices, it needs this [pr](https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pull/493) to be merged. If using another vdpa device plugin that has the same env variables with [Intel SR-IOV device plugin](https://github.com/intel/sriov-network-device-plugin) *(PCIDEVICE_<RESOUCE_NAME> and PCIDEVICE_<RESOUCE_NAME>_INFO)*, it should work well.

File: pkg/virt-launcher/virtwrap/converter/network.go (hunk 3 above)
Comment (makoto126) [reply pkg/virt-launcher/virtwrap/converter/network.go]: @prnaraya The resource loop should be outside the iface loop. Already fixed.

File: pkg/util/util.go (hunk 1 above)
Comment (makoto126) [reply pkg/util/util.go]: @prnaraya Already fixed.

File: pkg/virt-handler/vm.go [hunk 4]
```diff
@@ -566,6 +566,13 @@ func (d *VirtualMachineController) setupNetwork(vmi *v1.VirtualMachineInstance,
 				return neterrors.CreateCriticalNetworkError(fmt.Errorf("failed to set up tun device, %s", err))
 			}
 		}
+		vdpaDevs, _ := filepath.Glob(fmt.Sprintf("/proc/%d/root/dev/vhost-vdpa-*", isolationRes.Pid()))
```
Comment (stefano-garzarella) [inline pkg/virt-handler/vm.go:569]: Sorry, I don't know the code well, but I'm wondering, here how do you know it's a virtio-net device?

/dev/vhost-vdpa-* can be any virtio device emulated via vDPA.

Here is an example of 2 devices (block and net):
```bash
$ ls -l /dev/vhost-vdpa-*
crw-------. 1 root root 511, 0 Jul 14 10:55 /dev/vhost-vdpa-0
crw-------. 1 root root 511, 1 Jul 14 10:55 /dev/vhost-vdpa-1

$ vdpa dev
vdpa42: type block mgmtdev vdpasim_blk vendor_id 0 max_vqs 1 max_vq_size 256
vdpa13: type network mgmtdev vdpasim_net vendor_id 0 max_vqs 3 max_vq_size 256

$ ls /sys/devices/vdpa42/ | grep vhost
vhost-vdpa-0

$ cat /sys/devices/vdpa42/vhost-vdpa-*/dev
511:0

$ ls /sys/devices/vdpa13/ | grep vhost
vhost-vdpa-1

$ cat /sys/devices/vdpa13/vhost-vdpa-*/dev
511:1
```

Comment (makoto126) [reply pkg/virt-handler/vm.go:569]: @stefano-garzarella The vDPA devices in virt-launcher is allocated by device plugin, and there should be a configuration file that specifies the device type(1000). Here's my configuration file for the [Intel SR-IOV device plugin](https://github.com/intel/sriov-network-device-plugin):
```
# kubectl get cm -n kube-system sriovdp-config -o yaml
apiVersion: v1
data:
  config.json: |
    {
        "resourceList": [
            {
                "resourceName": "vdpa_jaguar_vhost",
                "selectors": [{
                    "vendors": ["1f53"],
                    "devices": ["1000"],
                    "drivers": ["jaguar"],
                    "vdpaType": "vhost"
                }]
            }
        ]
    }
```

Comment (makoto126) [reply pkg/virt-handler/vm.go:569]: @stefano-garzarella The line of code you mentioned looks for the vDPA device assigned by the device plugin and chown to qemu:qemu. If you don't do this, libvirt will tell you that you don't have enough privileges.

Comment (stefano-garzarella) [reply pkg/virt-handler/vm.go:569]: @makoto126 sorry, I don't know the code very well, but my concern is more about how to know if it's really the device we want. If we also add support for block devices in the future, should we change this code?

Comment (makoto126) [reply pkg/virt-handler/vm.go:569]: @stefano-garzarella If vDPA block support is added in the future, block devices also need to be chown to qemu:qemu, it shouldn't matter too much. If we must know whether the device is what we want, can read from /proc/%d/environ to get the same env variables with virt-launcher and parse the content, i don't think it's really necessary.

Comment (stefano-garzarella) [reply pkg/virt-handler/vm.go:569]: @makoto126 got it, thanks!

Comment (alicefr) [reply pkg/virt-handler/vm.go:569]: @makoto126 aren't you changing the permission of all devices `dev/vhost-vdpa-*` on the node in this way? What if there are vhost-vdpa devices that don't need to be assigned and used by KubeVirt?

Comment (stefano-garzarella) [reply pkg/virt-handler/vm.go:569]: @makoto126 another question: are we changing the permission of all `vhost-vdpa-*` devices on the node? If it is the case, is that safe?

Comment (makoto126) [reply pkg/virt-handler/vm.go:569]: @alicefr @stefano-garzarella No, only change the permission of all vhost-vdpa-* devices that assigned by the device plugin. 
For example, if the node has 5 vdpa devices, and the vmi declares 2 vdpa devices：
```
          interfaces:
          - masquerade: {}
            name: default
          - name: offload
            vdpa: {}
          - name: offload2
            vdpa: {}
      networks:
      - name: default
        pod: {}
      - multus:
          networkName: default/offload
        name: offload
      - multus:
          networkName: default/offload
        name: offload2
```
There will only be 2 vdpa devices in the virt-launcher container.

Comment (alicefr) [reply pkg/virt-handler/vm.go:569]: ah I see, you are using the container pid, thanks for the explanation

**Comment by xpivarc** [comment]
Hi @makoto126 ,
would you be able to submit a design document against https://github.com/kubevirt/community? 
Mainly I think it is important to cover what is required from a device plugin in order for Kubevirt to be able to consume the device.
/cc @EdDev 
another network binding.

**Comment by makoto126** [comment]
@xpivarc ok, i will submit a design document as soon as possible.

**Comment by EdDev** [comment]
> another network binding.

Indeed,

@makoto126 , we are working on exposing KubeVirt network bindings as plugins.
A design document [1] is already in place, taking shape gradually.
I would suggest you to feedback on it and see if the integration points there fit your needs.

[1] https://docs.google.com/document/d/13fLaqZuM8El_CT5XzxPv-1epKsB1-Spv_jtXveEkkdQ

**Comment by makoto126** [comment]
@EdDev I can't access it because I don't have permission.

**Comment by EdDev** [comment]
> @EdDev I can't access it because I don't have permission.

You need to join the `kubevirt-dev` googlegroup.
See https://github.com/kubevirt/kubevirt#community .

**Comment by makoto126** [comment]
@xpivarc https://github.com/kubevirt/community/pull/228

**Pull request closed without merging**